### PR TITLE
Add a provisional API for replacing the help formatter

### PR DESCRIPTION
**Pull request by mikeshardmind**
### Type

- [ ] Bugfix
- [X] Enhancement
- [X] New feature

### Description of the changes

- Adds an API for replacing the help formatter
- Defines which interfaces are required
- Adds some information for 3rd party developers
- Adds commands for resetting help behavior (to allow end users to fix issues with misbehaving cogs)

This is follow up from things which were planned originally in #2667 but tracked in #3601 and resolves a portion of that issue.

### CODE REVIEW

**Comment by mikeshardmind** [comment]
Something I considered this evening is that this might need to either wait until Red's formatter has support for sending help to a non-context based invoke (such as the occasional question of sending help on bot mention) or may need to be provisional so that the interface defined by the abc included is allowed to be expanded on with fewer restrictions.

**Comment by Jackenmen** [comment]
This needs to be shown in docs + I think we should mark this API as provisional for now. I'm gonna do the actual review (hopefully) soon™

**Comment by mikeshardmind** [comment]
I'll add that file to the docs with a note on it being provisional at some point today then.

File: docs/framework_commands.rst [hunk 1]
```diff
@@ -36,3 +36,17 @@ extend functionalities used throughout the bot, as outlined below.
     :no-undoc-members:
 
     .. autoclass:: APIToken
+
+==================
+Help Functionality
+==================
```
Comment (Jackenmen) [inline docs/framework_commands.rst]: Using top-level header here makes Sphinx show "Help functionality" as a separate top menu entry which I don't think it should?
image
```suggestion
******************
Help Functionality
******************
```

File: redbot/core/bot.py [hunk 2]
```diff
@@ -192,6 +189,52 @@ async def prefix_manager(bot, message) -> List[str]:
...
+        if type(self._help_formatter) is commands.help.RedHelpFormatter:
+            self._help_formatter = formatter
+        else:
+            raise RuntimeError("The formatter has already been overriden.")
+
+    def reset_help_formatter(self):
+        """
+        Resets Red's help formatter.
+
+        This exits for use in cog_unload for cogs which replace the formatter
```
Comment (Jackenmen) [inline redbot/core/bot.py]: typo and added backticks
```suggestion
        This exists for use in ``cog_unload`` for cogs which replace the formatter
```

File: redbot/core/commands/help.py [hunk 3]
```diff
@@ -1,18 +1,15 @@
...
 #
-# At a later date, there should be things added to support extra formatter
-# registration from 3rd party cogs.
-#
 # This exists due to deficiencies in discord.py which conflict
 # with our needs for per-context help settings
 # see https://github.com/Rapptz/discord.py/issues/2123
 #
-# While the issue above discusses this as theoretical, merely interacting with config within
+# While the issue above discusses this as theoretical,merely interacting with config within
```
Comment (Jackenmen) [inline redbot/core/commands/help.py]: add space back
```suggestion
# While the issue above discusses this as theoretical, merely interacting with config within
```

File: redbot/core/commands/help.py [hunk 4]
```diff
@@ -102,23 +105,72 @@ def __init__(self, *, last, not_found):
...
+        Red will not initialize a formatter for you,
+        and must be passed an initialized formatter.
+
+        If you want to use Red's existing settings, use ``HelpSettings.from_context``
+
+    .. warning::
+
+        This class is documented but provisional with expected changes.
+
+        In the future, this class will recieve changes to support
```
Comment (Jackenmen) [inline redbot/core/commands/help.py]: typo
```suggestion
        In the future, this class will receive changes to support
```

File: redbot/core/commands/help.py [hunk 5]
```diff
@@ -102,23 +105,72 @@ def __init__(self, *, last, not_found):
...
-    While currently, there is a global formatter, later plans include a context specific
-    formatter selector as well as an API for cogs to register/un-register a formatter with the bot.
+    While this exists as a class for easy partial overriding,
+    most implementations should not need or want a shared state.
+
+    .. warning::
 
-    When implementing your own formatter, at minimum you must provide an implementation of
-    `send_help` with identical signature.
+        This class is documented but may recieve changes between
```
Comment (Jackenmen) [inline redbot/core/commands/help.py]: typo
```suggestion
        This class is documented but may receive changes between
```

**Review by PredaaA — DISMISSED** [review]
I tested that PR on the functionnality, and everything is working well.

**Review by Jackenmen — COMMENTED** [review]
GH suggestion only review

**Review by Jackenmen — APPROVED** [review]
This PR works and since this is all marked as provisional, it can (and most likely will) still freely get any design changes later.

**Pull request merged**